substrate/frame/staking/src/pallet/mod.rs

```diff
@@ -1980,6 +1982,85 @@ pub mod pallet {
 
 			Ok(Some(T::WeightInfo::deprecate_controller_batch(controllers.len() as u32)).into())
 		}
+
+		/// Reset the state of a ledger which is in an inconsistent state.
+		///
+		/// The requirements to reset a ledger are the following:
+		/// * The stash is bonded;
+		/// * If the stash has an associated ledger, its state must be inconsistent;
+		/// * The reset ledger must become either a nominator or validator.
+		///
+		/// The `maybe_*` input parameters will overwrite the corresponding data and metadata of the
+		/// ledger associated with the stash. If the input parameters are not set, the ledger will
+		/// be reset with the current values.
+		///
+		/// Upon successful execution, this extrinsic will *recreate* the ledger based on its past
+		/// state and/or the `maybe_controller` and `maybe_total` input parameters.
+		#[pallet::call_index(29)]
+		#[pallet::weight(T::WeightInfo::reset_ledger())]
+		pub fn reset_ledger(
+			origin: OriginFor<T>,
+			stash: T::AccountId,
+			maybe_controller: Option<T::AccountId>,
+			maybe_total: Option<BalanceOf<T>>,
+		) -> DispatchResultWithPostInfo {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			let controller = Bonded::<T>::get(&stash).ok_or(Error::<T>::NotStash)?;
+
+			// ensure that this bond is in a bad state to proceed. i.e. one of two states: either
+			// the ledger for the controller does not exist or it exists but the stash is different
+			// than expected.
+			let maybe_other_ledger = Ledger::<T>::get(&controller);
+			ensure!(
+				maybe_other_ledger.as_ref().map(|l| l.stash != stash).unwrap_or(true),
+				Error::<T>::CannotResetLedger,
+			);
+
+			let current_lock =
+				<T::Currency as InspectLockableCurrency<T::AccountId>>::balance_locked(
+					crate::STAKING_ID,
+					&stash,
+				);
+
+			let new_total = if let Some(new_total) = maybe_total {
+				match maybe_other_ledger.clone() {
+					// if there is a ledger associated with this stash, it is the *wrong* ledger.
+					// Ensure that the "other" ledger has their locks/total updated so that the
+					// imbalance in the locks between ledgers is 0 (i.e. no more locks are
+					// created/deleted).
+					Some(mut other_ledger) => {
+						// note: ledger.total == ledger's staking lock amount.
+						other_ledger.total = new_total
```

Comment:
shouldn't we check what is the locked amount for the other ledger? And only modify ledger amount based on the lock. Current code might lead to other ledger not in sync anymore.

Reply:
Also it would probably be easier to read the code if we named all vars something like ledger_one, ledger_two, ledger_one_locked, ledger_one_stash and so on.

Reply:
> Also it would probably be easier to read the code if we named all vars something like ledger_one, ledger_two, ledger_one_locked, ledger_one_stash and so on.

I will use the following convention: `other_ledger` and `ledger`, `other_controller`, `controller`.

Reply:
> shouldn't we check what is the locked amount for the other ledger? And only modify ledger amount based on the lock. Current code might lead to other ledger not in sync anymore.

The code is assuming that the `maybe_total` input will set the correct amount to the current ledger (`StakingAdmin` has the responsibility to define what correct amount is). This code makes sure that the correct amount is set and that there's an imbalance of 0 between locks, ie, if other_ledger exists, decrease it's total and lock (in `other_ledger.update()`.

If the `maybe_total > other.ledger.lock` (note that `lock == ledger.amount`), then the extrinsic fails since we can't guarantee the imbalance and the call was wrong.

It seems to me that we don't need to check the other ledger lock since the other ledger is not corrupted and we thus we guarantee that lock == ledger.amount.

Reply:
Closing this thread as the code has considerably changed too.